### ISSUE TITLE
RuntimeError: invalid gradient at index 0 - expected shape [] but got [1]

### ISSUE THREAD

**Issue by gobbedy** [Zeleni9/pytorch-wgan#1]
Hello,

Thanks for posting your code, I'm finding it hard to find a WGAN-GP implementation in pytorch.

I've tried your implementation work but I'm completely stuck. I did get past the error below but got hit with many others, so I think my fix was wrong.

I'm running this command
main.py --model WGAN-GP --is_train True --download True --dataroot datasets/fashion-mnist --dataset fashion-mnist --epochs 30 --cuda True --batch_size 64

Traceback (most recent call last):
  File "C:/Users/Guillaume/PycharmProjects/pytorch-wgan2/main.py", line 41, in <module>
    main(args)
  File "C:/Users/Guillaume/PycharmProjects/pytorch-wgan2/main.py", line 30, in main
    model.train(train_loader)
  File "C:\Users\Guillaume\PycharmProjects\pytorch-wgan2\models\wgan_gradient_penalty.py", line 170, in train
    d_loss_real.backward(mone)
  File "C:\Program Files\Python36\lib\site-packages\torch\tensor.py", line 93, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "C:\Program Files\Python36\lib\site-packages\torch\autograd\__init__.py", line 90, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: invalid gradient at index 0 - expected shape [] but got [1]

Basically, I think it's some backward incompatibility with newer versions of pytorch (I'm using 0.4.1).

Do you have any idea how to fix? I'd be very grateful.

**Comment by Zeleni9** [comment]
Hi, thank you for stating this problem. I did not touch GAN code for a year, I will check it probably tomorrow, but the problem is that I don’t have GPU at the moment so I can’t really test it. Try it with pytorch 0.3, the error should not persist. Check out migration guide. Link: https://pytorch.org/blog/pytorch-0_4_0-migration-guide/   

Edit: Actually I found problem, there could be on two places maybe problems. Try changing part in code on line 134   one = torch.FloatTensor([1]) —>    one = torch.tensor(1.0)
If that does not solve the problem, then try changing these snippet of code:

gradients = autograd.grad(outputs = prob_interpolated, inputs = interpolated, grad_outputs = torch.ones(prob_interpolated.size()).cude(self.cuda_index) if self.cude else torch.ones(prob_interpolated.size()), create_graph = True, retain_graph = True)[0]

to 

gradients = autograd.grad(outputs = prob_interpolated, inputs = interpolated, grad_outputs = torch.[ones](prob_interpolated.size()).cude(self.cuda_index) if self.cude else  torch.[ones](prob_interpolated.size()), create_graph = True, retain_graph = True)[0]

Links: 
 https://stackoverflow.com/questions/52547157/pytorch-0-4-1-invalid-gradient-at-index-0-expected-shape-but-got-1
 https://github.com/ChenRocks/fast_abs_rl/issues/21  (try also torch.tensor(1.0) or torch.ones(1))

Something of these things should fix it. The code is working with stated requirements and version. 
Filip.
